src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,156 @@
 """Implementation of Rule L003."""
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, Iterable, List, Optional, Sequence, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+class _DictLikeDataCls:
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+    def __setitem__(self, key, item):
+        return setattr(self, key, item)
+
+    def items(self):
+        for field_el in dataclasses.fields(self):
+            key = field_el.name
+            yield key, getattr(self, key)
+
+
+@dataclasses.dataclass
+class _LineSummary(_DictLikeDataCls):
+    """A dataobject to represent a line.
+
+    Doubles as a memory object (holding state)
+    that can be converted to a fresh summary instance
+    """
+
+    line_no: int = 0
+    templated_line: Optional[int] = None
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    ignoreable_reference: List[bool] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_size: int = 1
+    indent_balance: int = 0
+    hanging_indent: Optional[int] = None
+    clean_indent: bool = True
+    template_content: str = ""
+    templated_line_type: Optional[str] = None
+    line_anchor: Optional[BaseSegment] = None
+
+    # These are only required for carrying state between lines
+    anchor_indent_balance: int = 0
```

Comment:
@barrywhart  - Im a little confused by what `anchor_indent_balance` and `indent_balance` are at which time in which roles.

The code is really just the same as it was, just that the state of `anchor_indent_balance` is saved also to this object.

I feel as though one of them is the indent balance at the end of the line (once fully processed) and one is the balance at the line at the point where the anchor was / is

The line we are calculating fixes for will always have the same, as with the current algo we break when we reach the first anchor of the last/current line

There are some weird inversions going on (line 124) and Id like to correct them any idea,

Reply:
I don't know. If you want to tag Alan, he may be the best able to answer questions about the existing code.